File: Lab12/Lab12.2/app/src/main/java/com/example/calendarnotes/MainActivity.java
package com.example.calendarnotes;


import android.app.Activity;
import android.app.AlertDialog;
import android.content.DialogInterface;
import android.os.Bundle;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ListView;
import android.widget.TextView;

import androidx.activity.EdgeToEdge;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;

public class MainActivity extends Activity {

    // Khai báo các biến
    ListView lv;
    ArrayList<String> arraywork;
    ArrayAdapter<String> arrAdapter;
    EditText edth, edtwork, edtm;
    TextView txtdate;
    Button btnwork;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        edth = findViewById(R.id.edthour);
        edtm = findViewById(R.id.edtm);
        edtwork = findViewById(R.id.edtwork);
        btnwork = findViewById(R.id.btnadd);
        lv = findViewById(R.id.listView1);
        txtdate = findViewById(R.id.txtdate);

        // Ở đây chúng ta không sử dụng dữ liệu cố định, mà dữ liệu của ListView được cập nhật
// trong quá trình chạy, nên ở đây ta tạo:
// Khai báo mảng ArrayList kiểu String rỗng
        arraywork = new ArrayList<>();

// Khai báo ArrayAdapter, đưa mảng dữ liệu vào ArrayAdapter
        arrAdapter = new ArrayAdapter<>(this, android.R.layout.simple_list_item_1, arraywork);

// Đưa Adapter có dữ liệu lên ListView
        lv.setAdapter(arrAdapter);

// Lấy ngày giờ hệ thống
        Date currentDate = Calendar.getInstance().getTime();

// Format theo định dạng dd/MM/yyyy
        java.text.SimpleDateFormat simpleDateFormat = new java.text.SimpleDateFormat("dd/MM/yyyy");

// Hiện thị lên TextView
        txtdate.setText("Hôm Nay: " + simpleDateFormat.format(currentDate));

// Viết phương thức khi Click vào Button btnwork
        btnwork.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                // TODO Auto-generated method stub

                // Nếu 1 trong 3 EditText không có nội dung thì hiện lên thông báo bằng hộp thoại Dialog
                if (edtwork.getText().toString().equals("") ||
                        edth.getText().toString().equals("") ||
                        edtm.getText().toString().equals("")) {

                    AlertDialog.Builder builder = new AlertDialog.Builder(MainActivity.this);
                    builder.setTitle("Info missing");
                    builder.setMessage("Please enter all information of the work");
                    builder.setPositiveButton("Continue", new DialogInterface.OnClickListener() {
                        public void onClick(DialogInterface dialog, int which) {
                            // TODO Auto-generated method stub
                        }
                    });
                    builder.show();
                }
                else {
                    String str = edtwork.getText().toString() + " - " +
                            edth.getText().toString() + ":" + edtm.getText().toString();

                    // Để thêm dữ liệu cho ListView, chúng ta cần 2 bước:
                    // Cập thêm liệu cho mảng
                    arraywork.add(str);         // Xóa: arraywork.remove(i)

                    // Cập nhật lại Adapter
                    arrAdapter.notifyDataSetChanged();

                    edth.setText("");
                    edtm.setText("");
                    edtwork.setText("");
                }
            }
        });

    }
}

File: Lab20/app/src/main/java/com/example/arsyntask/MainActivity.kt
package com.example.arsyntask

import android.os.AsyncTask
import android.os.Bundle
import android.os.SystemClock
import android.widget.Button
import android.widget.ProgressBar
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity

class MainActivity : AppCompatActivity() {

    private lateinit var btnStart: Button
    private lateinit var progressBar: ProgressBar
    private lateinit var txtMsg: TextView
    private var mytt: MyAsyncTask? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        btnStart = findViewById(R.id.button1)
        progressBar = findViewById(R.id.progressBar1)
        txtMsg = findViewById(R.id.textView)

        btnStart.setOnClickListener {
            doStart()
        }
    }

    private fun doStart() {
        mytt = MyAsyncTask(this)
        mytt?.execute()
    }

    // AsyncTask đã bị deprecated nhưng vẫn dùng cho mục đích học tập
    class MyAsyncTask(private val context: MainActivity) : AsyncTask<Void, Int, Void>() {

        override fun onPreExecute() {
            super.onPreExecute()
            Toast.makeText(context, "onPreExecute!", Toast.LENGTH_LONG).show()
        }

        override fun doInBackground(vararg params: Void?): Void? {
            for (i in 0..100) {
                SystemClock.sleep(100) // sleep 100ms
                publishProgress(i)
            }
            return null
        }

        override fun onProgressUpdate(vararg values: Int?) {
            super.onProgressUpdate(*values)
            val progress = values[0] ?: 0
            context.progressBar.progress = progress
            context.txtMsg.text = "$progress%"
        }

        override fun onPostExecute(result: Void?) {
            super.onPostExecute(result)
            Toast.makeText(context, "Update xong roi do!", Toast.LENGTH_LONG).show()
        }
    }
}